pkg/storage/utils/decomposedfs/upload.go

```diff
@@ -383,12 +397,58 @@ func (fs *Decomposedfs) GetUpload(ctx context.Context, id string) (tusd.Upload,
 	return &fileUpload{
 		info:     info,
 		binPath:  info.Storage["BinPath"],
-		infoPath: infoPath,
+		infoPath: filepath.Join(fs.o.Root, "uploads", id+".info"),
 		fs:       fs,
 		ctx:      ctx,
 	}, nil
 }
 
+// ListUploads returns a list of all incomplete uploads
+func (fs *Decomposedfs) ListUploads() ([]tusd.FileInfo, error) {
+	infos, err := filepath.Glob(filepath.Join(fs.o.Root, "uploads", "*.info"))
+	if err != nil {
+		return nil, err
+	}
+	list := make([]tusd.FileInfo, len(infos))
+	idRegexp := regexp.MustCompile(".*/([^/]+).info")
+	for i, info := range infos {
+		match := idRegexp.FindStringSubmatch(info)
+		upload, err := fs.readInfo(match[1])
+		if err != nil {
+			return nil, err
+		}
+		list[i] = upload
+	}
+	return list, nil
+}
+
+// PurgeExpiredUploads scans the fs for expired downloads and removes any leftovers
+func (fs *Decomposedfs) PurgeExpiredUploads(purgedChan chan<- tusd.FileInfo) error {
```

Comment:
Imho that should only purge an upload not list the uploads too. That is already done by `ListUploads`. One should call `ListUploads` then `PurgeUpload` when the upload expired. There is no need for `PurgeExpiredUploads` to copy the same logic as `ListUploads`

Reply:
Hm, `PurgeExpiredUploads` iterates over all uploads just like `ListUploads`, but it doesn't really copy its logic.
`ListUploads` returns a list of all the uploads it finds while `PurgeExpiredUploads` looks for uploads that are expired an deletes those. Does that make sense?

Reply:
Looking at the code again I will DRY it up some by extracting the list-compiling code into a separate function.